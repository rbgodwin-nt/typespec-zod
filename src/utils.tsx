import { createNamePolicy, NamePolicy, Refkey } from "@alloy-js/core";
import { Children } from "@alloy-js/core/jsx-runtime";
import {
  FunctionCallExpression,
  MemberExpression,
  TypeScriptElements,
} from "@alloy-js/typescript";
import { Program, Type } from "@typespec/compiler";
import { $ } from "@typespec/compiler/typekit";
import { camelCase, pascalCase } from "change-case";
import {
  getEmitOptionsForType,
  ZodOptionsContext,
} from "./context/zod-options.js";
import { zod } from "./external-packages/zod.js";

export const refkeySym = Symbol.for("@pavones/typespec-zod.refkey");

/**
 * Returns true if the given type is a declaration or an instantiation of a
 * declaration.
 */
export function isDeclaration(program: Program, type: Type): boolean {
  switch (type.kind) {
    case "Namespace":
    case "Interface":
    case "Operation":
    case "EnumMember":
      // TODO: this should reference the enum member via
      // target.enum.Name
      return false;
    case "UnionVariant":
      return false;

    case "Model":
      if (
        ($(program).array.is(type) || $(program).record.is(type)) &&
        isBuiltIn(program, type)
      ) {
        return false;
      }

      return Boolean(type.name);
    case "Union":
      return Boolean(type.name);
    case "Enum":
      return true;
    case "Scalar":
      return true;
    default:
      return false;
  }
}

// typekit doesn't consider things which have properties as records
// even though they are?
export function isRecord(program: Program, type: Type): boolean {
  return (
    type.kind === "Model" &&
    !!type.indexer &&
    type.indexer.key === $(program).builtin.string
  );
}

export function shouldReference(
  program: Program,
  type: Type,
  options?: ZodOptionsContext,
) {
  return (
    isDeclaration(program, type) &&
    !isBuiltIn(program, type) &&
    (!options ||
      !getEmitOptionsForType(program, type, options?.customEmit)?.noDeclaration)
  );
}

export function isBuiltIn(program: Program, type: Type) {
  if (type.kind === "ModelProperty" && type.model) {
    type = type.model;
  }

  if (!("namespace" in type) || type.namespace === undefined) {
    return false;
  }

  const globalNs = program.getGlobalNamespaceType();
  let tln = type.namespace;
  if (tln === globalNs) {
    return false;
  }

  while (tln.namespace !== globalNs) {
    tln = tln.namespace!;
  }

  return tln === globalNs.namespaces.get("TypeSpec");
}

/**
 * This API takes an array of types and returns those types in an array of
 * cyclesets. A cycleset is a group of types which together form a circular
 * reference. The array of cyclesets is ordered such that no type reachable from
 * a given cycleset is reachable from any cycleset earlier in the array. In
 * other words, if you emit the cyclesets in order there will be no forward
 * references.
 */
export function createCycleSets(types: Type[]): Type[][] {
  const inputTypes = new Set(types);

  let index = 0; // Unique index assigned to each node

  const stack: Type[] = [];
  const onStack = new Set<Type>();

  /* Map of type to the type's index */
  const indices = new Map<Type, number>();

  /* Map to store the smallest type index reachable from the given type */
  const lowlink = new Map<Type, number>();

  const sccs: Type[][] = [];

  for (const node of types) {
    if (!indices.has(node)) {
      strongConnect(node);
    }
  }

  return sccs;

  function referencedTypes(type: Type): Type[] {
    switch (type.kind) {
      case "Model":
        return [
          ...(type.baseModel ? [type.baseModel] : []),
          ...(type.indexer ? [type.indexer.key, type.indexer.value] : []),
          ...[...type.properties.values()].map((p) => p.type),
        ];

      case "Union":
        return [...type.variants.values()].map((v) =>
          v.kind === "UnionVariant" ? v.type : v,
        );
      case "UnionVariant":
        return [type.type];
      case "Interface":
        return [...type.operations.values()];
      case "Operation":
        return [type.parameters, type.returnType];
      case "Enum":
        return [];
      case "Scalar":
        return type.baseScalar ? [type.baseScalar] : [];
      case "Tuple":
        return type.values;
      case "Namespace":
        return [
          ...type.operations.values(),
          ...type.scalars.values(),
          ...type.models.values(),
          ...type.enums.values(),
          ...type.interfaces.values(),
          ...type.namespaces.values(),
        ];
      default:
        return [];
    }
  }

  // The main recursive function that implements Tarjan's algorithm.
  function strongConnect(v: Type): void {
    // Set the depth index for v to the smallest unused index
    indices.set(v, index);
    lowlink.set(v, index);
    index++;
    stack.push(v);
    onStack.add(v);

    // Consider successors of v
    for (const w of referencedTypes(v)) {
      if (!indices.has(w)) {
        // Successor w has not yet been visited; recurse on it.
        strongConnect(w);
        // After recursion, update lowlink[v]
        lowlink.set(v, Math.min(lowlink.get(v)!, lowlink.get(w)!));
      } else if (onStack.has(w)) {
        // If w is in the current SCC (i.e. on the stack), update lowlink[v]
        lowlink.set(v, Math.min(lowlink.get(v)!, indices.get(w)!));
      }
    }

    // If v is a root node, pop the stack and generate an SCC.
    if (lowlink.get(v) === indices.get(v)) {
      const component: Type[] = [];
      let w: Type;
      do {
        w = stack.pop()!;
        onStack.delete(w);
        component.push(w);
      } while (w !== v);

      const scc = component.filter((v) => inputTypes.has(v));
      if (scc.length > 0) {
        sccs.push(scc);
      }
    }
  }
}

export function call(target: string, ...args: Children[]) {
  return <FunctionCallExpression target={target} args={args} />;
}

export function memberExpr(...parts: Children[]) {
  return <MemberExpression children={parts} />;
}

export function zodMemberExpr(...parts: Children[]) {
  return memberExpr(refkeyPart(zod.z), ...parts);
}

export function idPart(id: string) {
  return <MemberExpression.Part id={id} />;
}

export function refkeyPart(refkey: Refkey) {
  return <MemberExpression.Part refkey={refkey} />;
}

export function callPart(target: string | Refkey, ...args: Children[]) {
  return (
    <MemberExpression>
      {typeof target === "string" ? idPart(target) : refkeyPart(target)}
      <MemberExpression.Part args={args} />;
    </MemberExpression>
  );
}

export function createZodNamePolicy(
  format: "pascal-case-schema" | "camel-case" | undefined,
): NamePolicy<TypeScriptElements> {
  const caseOptions = {
    prefixCharacters: "$_",
    suffixCharacters: "$_",
  };
  return createNamePolicy((name, element) => {
    if (format === "pascal-case-schema") {
      switch (element) {
        case "variable":
          return pascalCase(name, caseOptions) + "Schema";
        case "class":
        case "type":
        case "interface":
        case "enum":
        case "enum-member":
          return pascalCase(name, caseOptions);
        default:
          break;
      }
    }
    return camelCase(name, caseOptions);
  });
}

export const pascalZodNamePolicy = createZodNamePolicy("pascal-case-schema");
export const camelZodNamePolicy = createZodNamePolicy("camel-case");
