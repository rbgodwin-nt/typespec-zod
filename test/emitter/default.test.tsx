import { expect, it } from "vitest";
import { createEmitterTestRunner } from "../utils.jsx";


it("emits all declarations", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`
    model MyModel {
      id: string;
    }

    model MyModelArray is Array<MyModel> {}
    
    scalar MyScalar extends string;

    enum MyEnum {
      Value: 1
    }

    union MyFoo {
      1, 2
    }
  `);

  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});

it("handles references by doing a topological sort", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`
    union MyUnion {
      one: MyModelArray;
      two: MyModel;
    }

    model MyModelArray is Array<MyModel> {}

    model MyModel {
      id: string;
    }
  `);

  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});

it("handles the readme sample", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`

    model PetBase {
      age: uint8;
    
      @maxLength(20)
      name: string
    }
    
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);

  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});

it("doesn't emit things from built-in libraries", async () => {
  const runner = await createEmitterTestRunner({}, true);
  await runner.compile(`
    model PetBase {
      age: uint8;
    
      @maxLength(20)
      name: string;
    }

    model RefHttp {
      foo: OkResponse;
    }
  `);

  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});


it("handles the readme sample and emits PascalCaseSchema when naming-style is pascal-case-schema", async () => {
  const runner = await createEmitterTestRunner({ "naming-style": "pascal-case-schema" });
  
  await runner.compile(`
    model PetBase {
      age: uint8;
      @maxLength(20)
      name: string
    }
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);
  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");
  expect(text).toMatchSnapshot();
});

it("handles the readme sample and emits camelCase when nameing-style is camel-case", async () => {
  const runner = await createEmitterTestRunner({ "naming-style": "camel-case" });
  await runner.compile(`
    model PetBase {
      age: uint8;
      @maxLength(20)
      name: string
    }
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);
  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");
  expect(text).toMatchSnapshot();
});

it("emits Zod infer types when emitZodInfer is true", async () => {
  const runner = await createEmitterTestRunner({ emitZodInfer: true });
  await runner.compile(`
    model PetBase {
      age: uint8;
      @maxLength(20)
      name: string
    }
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);
  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");
  expect(text).toMatchSnapshot();
});

it("emits Zod infer types when emitZodInfer is true and naming-style is pascal-case-schema", async () => {
  const runner = await createEmitterTestRunner({ emitZodInfer: true, "naming-style": "pascal-case-schema" });
  await runner.compile(`
    model PetBase {
      age: uint8;
      @maxLength(20)
      name: string
    }
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);
  const { text } = await runner.program.host.readFile("@pavones/typespec-zod/models.ts");
  expect(text).toMatchSnapshot();
});

