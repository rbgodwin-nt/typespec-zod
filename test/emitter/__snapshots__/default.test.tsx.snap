// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`doesn't emit things from built-in libraries 1`] = `
"import { z } from "zod";

export const petBase = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export const refHttp = z.object({
  foo: z.object({
    statusCode: z.literal(200),
  }),
});"
`;

exports[`emits Zod infer types when emitZodInfer is true 1`] = `
"import { z } from "zod";

export const petBase = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export type PetBase = z.infer<typeof petBase>;

export const dog = petBase.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export type Dog = z.infer<typeof dog>;

export const cat = petBase.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615n),
}));

export type Cat = z.infer<typeof cat>;

export const pet = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: dog,
    }),
    z.object({
      kind: z.literal("cat"),
      value: cat,
    })
  ]
);
export type Pet = z.infer<typeof pet>;
"`;

exports[`emits Zod infer types when emitZodInfer is true and naming-style is pascal-case-schema 1`] = `
"import { z } from "zod";

export const PetBaseSchema = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export type PetBase = z.infer<typeof PetBaseSchema>;

export const DogSchema = PetBaseSchema.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export type Dog = z.infer<typeof DogSchema>;

export const CatSchema = PetBaseSchema.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615n),
}));

export type Cat = z.infer<typeof CatSchema>;

export const PetSchema = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: DogSchema,
    }),
    z.object({
      kind: z.literal("cat"),
      value: CatSchema,
    })
  ]
);

export type Pet = z.infer<typeof PetSchema>;
"
`;

exports[`emits all declarations 1`] = `
"import { z } from "zod";

export const myModel = z.object({
  id: z.string(),
});

export const myModelArray = z.array(myModel);

export const myScalar = z.string();

export const myFoo = z.union([z.literal(1), z.literal(2)]);

export const myEnum = z.enum([1]);"
`;

exports[`handles references by doing a topological sort 1`] = `
"import { z } from "zod";

export const myModel = z.object({
  id: z.string(),
});

export const myModelArray = z.array(myModel);

export const myUnion = z.union([myModelArray, myModel]);"
`;

exports[`handles the readme sample 1`] = `
"import { z } from "zod";

export const petBase = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export const dog = petBase.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export const cat = petBase.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615n),
}));

export const pet = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: dog,
    }),
    z.object({
      kind: z.literal("cat"),
      value: cat,
    })
  ]
);"
`;

exports[`handles the readme sample and emits PascalCaseSchema when naming-style is pascal-case-schema 1`] = `
"import { z } from "zod";

export const PetBaseSchema = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export const DogSchema = PetBaseSchema.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export const CatSchema = PetBaseSchema.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615n),
}));

export const PetSchema = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: DogSchema,
    }),
    z.object({
      kind: z.literal("cat"),
      value: CatSchema,
    })
  ]
);"
`;

exports[`handles the readme sample and emits camelCase when naming-style is camel-case 1`] = `
"import { z } from "zod";

export const petBase = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(20),
});

export const dog = petBase.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export const cat = petBase.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615n),
}));

export const pet = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: dog,
    }),
    z.object({
      kind: z.literal("cat"),
      value: cat,
    })
  ]
);"
`;
